### PR TITLE
fix deadlock in hackney_pool during request timeout

### PR DESCRIPTION
**Pull request by jimdigriz**
This mostly resolves https://github.com/benoitc/hackney/issues/420 (and also includes https://github.com/benoitc/hackney/pull/419).

A race condition still exists, between the request timeout occurring till when the cast to `checkout_cancel` is processed.  I suspect it is really unlikely, though fortunately it is detectable so we could choose to do something about it later if this occurs.

A possible improvement would be to when the race occurs we dig around the `client` state element and reclaim the socket back into the pool?

### CODE REVIEW

**Comment by Hinidu** [comment]
@jimdigriz Did you see that it fails some test?

**Comment by jimdigriz** [comment]
So does the master branch

**Comment by Hinidu** [comment]
> So does the master branch

Hmmm, no, it fails not the test you've added but another one. Failed test is passed on master. See [Travis CI report](https://travis-ci.org/benoitc/hackney/builds/286401503?utm_source=github_status&utm_medium=notification).

File: src/hackney_pool.erl [hunk 1]
```diff
@@ -304,6 +307,12 @@ handle_cast({set_maxconn, MaxConn}, State) ->
 handle_cast({set_timeout, NewTimeout}, State) ->
   {noreply, State#state{timeout=NewTimeout}};
 
+handle_cast({checkout_cancel, Dest, RequestRef}, State) ->
```
Comment (benoitc) [inline src/hackney_pool.erl]: How much should it be different from a 'DOWN' case ?
https://github.com/benoitc/hackney/blob/master/src/hackney_pool.erl

imo we should go back in the queue and trigger another waiting client in such case. Thoughts?

**Comment by benoitc** [comment]
@jimdigriz mmm that tests seems related to the addition of the timeout error. Maybe a new edge case?

**Comment by jimdigriz** [comment]
I'll try to have a nosey at this tonight.

**Comment by benoitc** [comment]
@jimdigriz cool i will be available on irc tonight as well if you want to discuss some parts.

File: src/hackney_pool.erl (hunk 1 above)
Comment (jimdigriz) [reply src/hackney_pool.erl]: Sounds like a good idea to me, reworking...

**Comment by jimdigriz** [comment]
> >  So does the master branch

> Mmmm, no, it fails not the test you've added but another one. Failed test is passed on master. See Travis CI report.

The problem is with the unit test `hackney_pool_tests.erl`, it sets the pool connection size to 1 via the environment but never restores it...it should probably not be playing with the application environment anyway.

[Fixed the unit test](https://github.com/jimdigriz/hackney/commit/8032e2b0eaf3880fb436b0d09f23beced9690471).

**Comment by jimdigriz** [comment]
> i like the idea. I think this should be handle like a 'DOWN' case though, so it triggers another waiting client.

I think this is causing the gen_server to do extra work when we know `queue_out()` will always return `empty`; as we are cancelling a queued `checkout`.

For example we have `MaxConn` set to 10, we try to queue 11, it times out and we send in a cancellation.  Meanwhile `length(clients)=10, length(queue)=1`.  We sending in `checkout_cancel` and we will still have `length(clients)=10` as we are cancelling the request that is in `queue`.

Does this sound right?

**Comment by benoitc** [comment]
@jimdigriz mmm right you need to clean yourself, but I think we still need a way to trigger another waiting connection. Ie since you reduce the number of connections to be be released, let's handle one another from the queue and timeout ourself. Does it make sense?

**Comment by jimdigriz** [comment]
We are only removing requests from the `queue` dictionary, all of those are just awaiting for an available socket.  If a trigger could do something, it would have already from another socket, right?

What we can do though is fix the race by recovering the socket if we see that the request has moved from the `queue` into the `clients` dictionary.  Here we can see if someone else may be interested in the socket.

**Comment by jimdigriz** [comment]
What actually calls `DOWN`?  I am unable to find any `monitor()`/`link()`/`process_flag(trap_exit, true)` calls in `hackney_pool.erl`.

**Comment by benoitc** [comment]
@jimdigriz this is the manager: https://github.com/benoitc/hackney/blob/master/src/hackney_manager.erl#L408

This part is inherited from the past. The manager is responsible of maintaining the requests. This is the part i expect to replace soon by maintainting each connection states in their own process instead (ie generalising the async case). Which would also solve #439 .

I was thinking this cancel was doing exactly what you do right now somehow but preemptively . The reason to trigger (queuing out) after having deleted yourself is to enable another waiter to receive a socket. If not it will wait until another socket is released. Isn't it?

**Comment by jimdigriz** [comment]
Okay, figured out `no_socket` actually triggers the creation of a socket, no longer panic'ing :)

File: src/hackney_pool.erl [hunk 2]
```diff
@@ -304,6 +307,16 @@ handle_cast({set_maxconn, MaxConn}, State) ->
 handle_cast({set_timeout, NewTimeout}, State) ->
   {noreply, State#state{timeout=NewTimeout}};
 
+handle_cast({checkout_cancel, Dest, Ref}, State) ->
+  {Queues, Removed} = del_from_queue(Dest, Ref, State#state.queues),
+  case Removed of
+    true ->
+      NbWaiters = State#state.nb_waiters - 1,
+      {noreply, State#state{queues=Queues, nb_waiters=NbWaiters}};
```
Comment (benoitc) [inline src/hackney_pool.erl]: why not dequeuing it there too?

Comment (jimdigriz) [reply src/hackney_pool.erl]: As we have non-zero `nb_waiters` only want `length(clients)==MaxConns` right, otherwise when one of those clients finished, `hackney_pool` would have dequeued already?

Comment (benoitc) [reply src/hackney_pool.erl]: releasing a waiter only happen when a connection is put back in the loop or when a request is canceled: forcefully closed via  hackney:close/1 or when a process actually requesting crash. I think we are in the second case there, so imo we should dequeue one of the waiter to actually tell him to create a new request imo.

Comment (jimdigriz) [reply src/hackney_pool.erl]: We will exceed `pool_size` though?

Comment (benoitc) [reply src/hackney_pool.erl]: @jimdigriz right. applying your change now. Will release sometimes today :)
